lib/pg/connection.rb

```diff
@@ -268,5 +268,24 @@ def ssl_attributes
 		end
 	end
 
+	REDIRECT_METHODS = {
+		:exec => [:async_exec, :sync_exec],
+		:query => [:async_exec, :sync_exec],
+		:exec_params => [:async_exec, :sync_exec_params],
+		:prepare => [:async_prepare, :sync_prepare],
+		:exec_prepared => [:async_query_prepared, :sync_exec_prepared],
+		:describe_portal => [:async_describe_portal, :sync_describe_portal],
+		:describe_prepared => [:async_describe_prepared, :sync_describe_prepared],
+	}
+
+	def self.async_api=(enable)
+		REDIRECT_METHODS.each do |ali, (async, sync)|
+			remove_method(ali) if method_defined?(ali)
+			alias_method( ali, enable ? async : sync )
+		end
+	end
+
+	# pg-1.1.0+ defaults to libpq's async API for query related blocking methods
+	self.async_api = true
```

Comment:
Wouldn't this require bumping the major version? While this isn't a change in the method signature, it definitely changes the behavior of methods with the same name. I think this is a good change going forward, but I think it's significant enough to warrant a major-version change.

Reply:
> it definitely changes the behavior of methods

Does it? My argument was precisely that it doesn't. It changes the _implementation_, but not any ruby- (or postgres-) observable behaviour.

Reply:
I understand it as a bug, if the methods change their behavior due to this PR. That's why I enabled the test suite to run all specs through both sync and async versions of the methods.

There are two (known) differences between sync and async:

1. Signals are no longer delayed until the method returns.
2. The trace outputs are slightly different.

I would argue that delaying signals is a bug, not a feature, so that I think 1. is a fix rather than a behavioral change.

The trace output is an implementation specific feature. We already had a change between PostgreSQL 8.4 and 9.0, so that I count that as internal only. In fact this is the only behavioral change I did to the specs to make it pass.

We aren't in hurry with this change. So I still would like to test pg in various contexts before releasing, to make sure or test suite isn't fooling us. But in the end I would release it as pg-1.1.0.

Reply:
I guess it depends on what you consider the "API" with regards to semantic versioning. Don't misunderstand me: this change is unquestionably making the library better, and I would advocate for merging it even if it meant a major version change. I agree that the trace output is only revealing internal details which shouldn't have been counted on. I agree that delaying signals is not a feature, but it still is behavior which may change how some systems run. It's definitely on the fringes of what I'd consider a library's "API", but I thought the question still worth asking. If you both are unconcerned, however, I'm fine with just doing careful testing and releasing it as 1.1.0.